File: src/main/java/uk/yermak/audiobookconverter/AudioBookInfo.java
package uk.yermak.audiobookconverter;

import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import org.apache.commons.lang3.StringUtils;
import uk.yermak.audiobookconverter.fx.util.SmartIntegerProperty;
import uk.yermak.audiobookconverter.fx.util.SmartStringProperty;

import java.util.Collections;
import java.util.Map;

import static org.apache.commons.lang3.StringUtils.firstNonBlank;
import static org.apache.commons.lang3.StringUtils.trimToEmpty;


public record AudioBookInfo(SmartStringProperty title, SmartStringProperty writer,
                            SmartStringProperty narrator, SmartStringProperty series,
                            SmartStringProperty genre, SmartStringProperty year,
                            SmartIntegerProperty bookNumber, SmartIntegerProperty totalTracks,
                            SmartStringProperty comment,
                            ObservableList<Track> tracks) {


    public static AudioBookInfo instance(Map<String, String> tags) {
        int trackNumber = tags.get("track number") == null ? 0 : Integer.parseInt(tags.get("track number"));
        int trackCount = tags.get("track count") == null ? 0 : Integer.parseInt(tags.get("track count"));

        String track = tags.get("track");
        if (StringUtils.isNotEmpty(track)) {
            String[] split = StringUtils.split(track, "/");
            trackNumber = Integer.parseInt(split[0]);
            if (split.length > 1) {
                trackCount = Integer.parseInt(split[1]);
            }
        }

        return new AudioBookInfo(
                new SmartStringProperty(trimToEmpty(tags.get("title"))),
                new SmartStringProperty(trimToEmpty(tags.get("artist"))),
                new SmartStringProperty(firstNonBlank(trimToEmpty(tags.get("narratedby")), trimToEmpty(tags.get("composer")))),
                new SmartStringProperty(trimToEmpty(tags.get("album"))),
                new SmartStringProperty(trimToEmpty(tags.get("genre"))),
                new SmartStringProperty(firstNonBlank(trimToEmpty(tags.get("year")), StringUtils.substring(trimToEmpty(tags.get("date")), 0, 4))),
                new SmartIntegerProperty(trackNumber),
                new SmartIntegerProperty(trackCount),
                new SmartStringProperty(firstNonBlank(trimToEmpty(tags.get("comment")), trimToEmpty(tags.get("comment-0")))),
                FXCollections.observableArrayList()
        );
    }

    public static AudioBookInfo instance() {
        return instance(Collections.emptyMap());
    }
}

        

File: src/main/scala/uk/yermak/audiobookconverter/AudioBookInfo.scala
package uk.yermak.audiobookconverter

import javafx.collections.ObservableList

import scala.collection.mutable

class AudioBookInfo() {
  private var writer = ""
  private var narrator = ""
  private var title = ""
  private var series = ""
  private var genre = ""
  private var year = ""
  private var bookNumber = 0
  private var totalTracks = 0
  private var comment = ""
  private var longDescription = ""
  private var posters: ObservableList[ArtWork] = _

  def getSeries: String = {
    if (series == null) return title
    this.series
  }

  def setSeries(series: String): Unit = this.series = series

  def getWriter: String = this.writer

  def setWriter(writer: String): Unit = this.writer = writer

  def getComment: String = this.comment

  def setComment(comment: String): Unit = this.comment = comment

  def getGenre: String = this.genre

  def setGenre(genre: String): Unit = this.genre = genre

  def getTitle: String = this.title

  def setTitle(title: String): Unit = this.title = title

  def getBookNumber: Int = this.bookNumber

  def setBookNumber(bookNumber: Int): Unit = this.bookNumber = bookNumber

  def getNarrator: String = this.narrator

  def setNarrator(narrator: String): Unit = this.narrator = narrator

  def getYear: String = this.year

  def setYear(year: String): Unit = this.year = year

  def getTotalTracks: Int = totalTracks

  def setTotalTracks(totalTracks: Int): Unit = this.totalTracks = totalTracks

  def getLongDescription: String = longDescription

  def setLongDescription(longDescription: String): Unit =
    this.longDescription = longDescription

  def getPosters: ObservableList[ArtWork] = posters

  def setPosters(posters: ObservableList[ArtWork]): Unit =
    this.posters = posters

}

object AudioBookInfo {

  def instance(tags: mutable.Map[String, String]): AudioBookInfo = {
    val audioBookInfo = new AudioBookInfo
    if (tags != null) {
      audioBookInfo.setTitle(tags("title"))
      audioBookInfo.setWriter(tags("artist"))
      audioBookInfo.setNarrator(tags("album_artist"))
      audioBookInfo.setSeries(tags("album"))
      audioBookInfo.setYear(tags("year"))
      audioBookInfo.setComment(tags("comment-0"))
      audioBookInfo.setGenre(tags("genre"))
      val trackNumber = tags("track number")
      if (trackNumber != null)
        audioBookInfo.setBookNumber(trackNumber.asInstanceOf[Int])
      val trackCount = tags("track count")
      if (trackCount != null)
        audioBookInfo.setTotalTracks(trackCount.asInstanceOf[Int])
    }
    audioBookInfo
  }
}
